### PR TITLE
[llvm] Make sure OP_LSHL_IMM is transformed to a 64 bit shift even if the input is 32 bit, which can happen because the JIT IR is not regular. Fixes bug-17590.exe on arm64+fullaot+llvm.

### PR DESCRIPTION
**Pull request by vargaz**
Fixes https://github.com/mono/mono/issues/10527

### CODE REVIEW

**Comment by lewurm** [comment]
not sure if related:
```
MONO_PATH=/home/builder/jenkins/workspace/test-mono-pull-request-arm64-fullaot+llvm/mcs/class/lib/testing_aot_full ../../runtime/mono-wrapper --runtime=mobile -O=gsharedvt --llvm --aot=full,nimt-trampolines=2000,ntrampolines=10000,nrgctx-fetch-trampolines=256,ngsharedvt-trampolines=4000 generic-constrained.2.exe
[...]
#3  0x0000aaaae7fe2734 in mono_handle_native_crash (signal=signal@entry=0xaaaae85eae70 "SIGSEGV", ctx=ctx@entry=0xffffdba544f0, info=info@entry=0xffffdba54470) at mini-exceptions.c:3000
#4  0x0000aaaae7f76704 in mono_sigsegv_signal_handler (_dummy=11, _info=0xffffdba54470, context=0xffffdba544f0) at mini-runtime.c:3424
#5  <signal handler called>
#6  0x0000aaaae84fe7f4 in llvm::Value::getContext() const ()
#7  0x0000aaaae84fefa0 in llvm::Value::setNameImpl(llvm::Twine const&) ()
#8  0x0000aaaae84ff21c in llvm::Value::setName(llvm::Twine const&) ()
#9  0x0000aaaae8453d84 in LLVMBuildExtractValue ()
#10 0x0000aaaae800e1b0 in emit_entry_bb (ctx=ctx@entry=0xaaab0352dcd0, builder=builder@entry=0xaaab03525720) at mini-llvm.c:3040
#11 0x0000aaaae8011d70 in emit_method_inner (ctx=ctx@entry=0xaaab0352dcd0) at mini-llvm.c:7390
#12 0x0000aaaae8018fa0 in mono_llvm_emit_method (cfg=cfg@entry=0xaaab0352d640) at mini-llvm.c:7064
#13 0x0000aaaae8036d50 in mini_method_compile (method=method@entry=0xaaab0351a1e0, opts=opts@entry=391195135, domain=<optimized out>, flags=flags@entry=(JIT_FLAG_AOT | JIT_FLAG_FULL_AOT | JIT_FLAG_LLVM), parts=parts@entry=0, aot_method_index=aot_method_index@entry=1) at mini.c:3821
#14 0x0000aaaae7fcb608 in compile_method (acfg=acfg@entry=0xaaab0350ea80, method=0xaaab0351a1e0) at aot-compiler.c:7973
#15 0x0000aaaae7fcdc68 in compile_method (method=<optimized out>, acfg=0xaaab0350ea80) at aot-compiler.c:11250
```
from https://jenkins.mono-project.com/job/test-mono-pull-request-arm64-fullaot+llvm/71

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by vargaz** [comment]
That failure on arm64+llvm+aot is reproducible.

**Comment by lewurm** [comment]
yes, the `Linux AArch64 FullAOT+LLVM` failure looks related to this PR. I don't see it on `master`.

File: mono/mini/mini-llvm.c [hunk 1]
```diff
@@ -3029,7 +3029,8 @@ emit_entry_bb (EmitContext *ctx, LLVMBuilderRef builder)
 			ctx->addresses [reg] = build_alloca (ctx, ainfo->type);
 
 			size = mono_class_value_size (mono_class_from_mono_type (ainfo->type), NULL);
-			if (size < TARGET_SIZEOF_VOID_P) {
+			if (size == 0) {
+			} else if (size < TARGET_SIZEOF_VOID_P) {
```
Comment (jaykrell) [inline mono/mini/mini-llvm.c]: Imho, cosmetic, should be if `(size && size < TARGET_SIZEOF_VOID_P)`.

**Comment by vargaz** [comment]
@monojenkins build failed

**Comment by lewurm** [comment]
@monojenkins squash

**Comment by luhenry** [comment]
@monojenkins build Linux AArch64 FullAOT+LLVM

**Comment by akoeplinger** [comment]
@monojenkins backport 2018-08

**Pull request merged**